File: solutions/ts/21.ts
function treeHeight(
  tree: { value: string; left: any; right: any } | null
): number {
  if (!tree) return 0;
  return 1 + Math.max(treeHeight(tree.left), treeHeight(tree.right));
}

const giftTree = {
  value: "🎁",
  left: {
    value: "🎄",
    left: {
      value: "⭐",
      left: null,
      right: null,
    },
    right: {
      value: "🎅",
      left: null,
      right: null,
    },
  },
  right: {
    value: "❄️",
    left: null,
    right: {
      value: "🦌",
      left: null,
      right: null,
    },
  },
};

// Graphical representation of the tree:
//        🎁
//       /   \
//     🎄     ❄️
//    /  \      \
//  ⭐   🎅      🦌

// Function call
console.log(treeHeight(giftTree));
// Returns: 3

/* 
  Score: ★★★★★
*/

File: solutions/python/21.py
def tree_height(tree):
    if (not isinstance(tree, dict)):
        return 0
    left = tree_height(tree.get("left"))
    right = tree_height(tree.get("right"))
    return 1+max(left, right)


if __name__ == "__main__":
    gift_tree = {
        "value": "🎁",
        "left": {
            "value": "🎄",
            "left": {
                "value": "⭐",
                "left": None,
                "right": None,
            },
            "right": {
                "value": "🎅",
                "left": None,
                "right": None,
            },
        },
        "right": {
            "value": "❄️",
            "left": None,
            "right": {
                "value": "🦌",
                "left": None,
                "right": None,
            },
        },
    }

    # Graphical representation of the tree:
    #        🎁
    #       /   \
    #     🎄     ❄️
    #    /  \      \
    #  ⭐   🎅      🦌

    # Function call
    print(tree_height(gift_tree))
    # Returns: 3

"""
  Score: ★★★★★
"""
